### DISCUSSION TITLE
Some confusions with TryGetXXX

### DISCUSSION THREAD

**Discussion by michaelmairegger** [quartznet/quartznet#2122] [Ideas]
GetXXX and TryGetXXX is a bit confusing. E.g. GetInt with a non existent key returns 0. I expect to get an exception. Same for TryGetIntValue. The following branch is never hit:

    if (!context.MergedJobDataMap.TryGetIntValue("key that is not in job data map", out var lastValue))
    {
        // will never reach here
    }

the only solution that works is by using TryGetValue

I would suggest the following changes in behavior of the methods:

- Throw KeyNotFoundException on GetXXX if the key does not exist (actual behavior: returns default value)
- Return false on TryGetXXX and TryGetXXXValue if the key does not exists.

**Comment by jafin** [comment]
@michaelmairegger I think this is a bug.  

`TryGetIntValue` calls into `GetInt`.

the call

```c#

public virtual int GetInt(string key)
{
    var obj = this[key];

    try
    {
        return Convert.ToInt32(obj);  //returns 0 on null obj, probably not the expected intent.
    }
    catch (Exception)
    {
        throw new InvalidCastException("Identified object is not an Integer.");
    }
}
```

whereas it should be throwing an exception, which should bubble up and get caught and return false for TryGetIntValue

**Reply by michaelmairegger** [reply]
Exactly. I would expect GetXXX to throw a KeyNotFoundException if the key is not in the map. Furthermore TryGetXXXValue and TryGetXXX should return false if the key does not exist
Looking at the code the indexer this[key] of DirtyFlagMap<TKey, TValue> will not throw an exception neither. Wheras it sould as it inherits from IDictionary.

    context.MergedJobDataMap["key1"] = null;
    if (context.MergedJobDataMap["key1"] == context.MergedJobDataMap["key2"])
    {
        //this condition should not met
    }

But changing the code to this behavior, would be a breaking change.

**Reply by jafin** [reply]
I threw up a PR for the first issue. https://github.com/quartznet/quartznet/pull/2125    targeting 3.x
